src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogAgentClient.java

```diff
@@ -320,6 +481,32 @@ private boolean reinitializeAgentHttpClient(boolean force) {
         }
     }
 
+    protected boolean checkEvpProxySupportAndUpdateLogic() {
+        if (evpProxySupported) {
+            return true; // Once we have seen an Agent that supports EVP Proxy, we never check again.
+        }
+        if (System.currentTimeMillis() < (lastEvpProxyCheckTimeMs + EVP_PROXY_SUPPORT_TIME_BETWEEN_CHECKS_MS)) {
+            return evpProxySupported; // Wait at least 1 hour between checks, return the cached value
+        }
+        synchronized (DatadogAgentClient.class) {
+            if (!evpProxySupported) {
+                Set<String> supportedAgentEndpoints = fetchAgentSupportedEndpoints();
+                evpProxySupported = supportedAgentEndpoints.contains("/evp_proxy/v3/");
+                lastEvpProxyCheckTimeMs = System.currentTimeMillis();
+                if (evpProxySupported) {
+                    logger.fine("EVP Proxy is supported by the Agent.");
+                    traceBuildLogic = new DatadogWebhookBuildLogic(this);
+                    tracePipelineLogic = new DatadogWebhookPipelineLogic(this);
```

Comment:
Could we have race conditions here? Like, the logic started using the version with APM, and then, the logic changes, and it changes to WebHooks? Do we need to protect this case?

Reply:
If some spans get sent as traces and others as webhooks, nothing bad happens. I just checked it and the backend even displays them as the same pipeline because of the same trace id!